### PR TITLE
[DSIP-13][python]Sub problem: gitlab, OSS and S3 resource plug-ins

### PR DESCRIPTION
**Pull request by xdu-chenrj**
## Purpose of the pull request

This pull request adds a `GitLab`,  `OSS`, `S3` resource plug-in feature #10911 

mail-list
https://lists.apache.org/thread/r1f9wwlrj6jrnyc1hwpf8qoq8zr8n7j1

## Brief change log

     Added GitLab,OSS and S3 resource plug-ins
     Added testing of GitLab,OSS and S3 resource plug-ins
     Added documentation on how to use GitLab,OSS and S3 resource plug-ins

## Verify this pull request

This change added tests and can be verified as follows:

     Added testing of GitLab, OSS, S3 resource plug-ins
     Manually verified the change by testing locally

### CODE REVIEW

**Comment by zhongjiajie** [comment]
@xdu-chenrj it is seem conflict here, can you resolve it ?

**Comment by xdu-chenrj** [comment]
> @xdu-chenrj it is seem conflict here, can you resolve it ?

yeah

File: dolphinscheduler-python/pydolphinscheduler/tests/resources_plugin/test_s3.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+@pytest.mark.skip(reason="This test requires s3 services")
+@pytest.mark.parametrize(
+    "attr, expected",
+    [
+        (
+            {
+                "init": {
+                    "prefix": "https://ds-resource-plugin-private.s3.amazonaws.com/dir/",
+                    "access_key_id": "LTAI5tP25Mxx",
+                    "access_key_secret": "cSur23Qbxx",
```
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/tests/resources_plugin/test_s3.py:59]: Is this the true access key and secret, or just mark the ending?

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    """OSS object, declare OSS resource plugin for task and workflow to dolphinscheduler.
+
+    :param prefix: A string representing the prefix of OSS.
+    :param access_key_id: A string representing the ID of AccessKey for AliCloud OSS
+
+    to access private files.
+
+    :param access_key_secret: A string representing the secret of AccessKey for AliCloud OSS
+
+    to access private files.
```
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py]:
```suggestion
    :param access_key_id: A string representing the ID of AccessKey for AliCloud OSS to access private files.

    :param access_key_secret: A string representing the secret of AccessKey for AliCloud OSS to access private files.
```

Comment (zhongjiajie) [reply dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py]: please remove the unnecessary blank line

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/s3.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    """S3 object, declare S3 resource plugin for task and workflow to dolphinscheduler.
+
+    :param prefix: A string representing the prefix of S3.
+    :param access_key_id: A string representing the ID of AccessKey for Amazon S3
+
+    to access private files.
+
+    :param access_key_secret: A string representing the secret of AccessKey for Amazon S3
+
+    to access private files.
```
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/s3.py]: remove the unnecessary blank line here too

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/gitlab.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                host_end = self.get_index(path, "/", 3)
+                self._git_file_info = GitLabFileInfo(
+                    host=path[0:host_end],
+                    repo_name=elements[i],
+                    branch=elements[i + 3],
+                    file_path="/".join(
+                        str(elements[j]) for j in range(i + 4, len(elements))
+                    ),
+                    user="/".join(str(elements[j]) for j in range(3, i)),
+                )
```
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/gitlab.py]: can we use other way to handle this, I think we use too many constant in this method

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    def get_bucket_file_info(self, path: str):
+        """Get file information from the file url, like repository name, user, branch, and file path."""
+        elements = path.split("/")
+        self.get_index(path, "/", 3)
+        self._bucket_file_info = OSSFileInfo(
+            endpoint="https://" + elements[2].split(".")[1] + ".aliyuncs.com",
+            bucket=elements[2].split(".")[0],
+            file_path="/".join(str(elements[i]) for i in range(3, len(elements))),
+        )
```
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py]: same here, we should not use too many bare constant here

File: dolphinscheduler-python/pydolphinscheduler/tests/resources_plugin/test_gitlab.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import pytest
+
+from pydolphinscheduler.resources_plugin.gitlab import GitLab
+
+
+@pytest.mark.parametrize(
+    "attr, expected",
+    [
+        (
+            "https://gitlab.com/chenruijie/ds-gitlab/-/blob/main/union.sh",
```
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/tests/resources_plugin/test_gitlab.py]: can we remove all author-related message `chenruijie` to other like `pydolphinscheduler` or some thing else?
```suggestion
            "https://gitlab.com/pydolphinscheduler/ds-gitlab/-/blob/main/union.sh",
```

Comment (xdu-chenrj) [reply dolphinscheduler-python/pydolphinscheduler/tests/resources_plugin/test_gitlab.py]: > can we remove all author-related message `chenruijie` to other like `pydolphinscheduler` or some thing else?

i know

File: dolphinscheduler-python/pydolphinscheduler/tests/resources_plugin/test_s3.py (hunk 1 above)
Comment (xdu-chenrj) [reply dolphinscheduler-python/pydolphinscheduler/tests/resources_plugin/test_s3.py:59]: mark the ending

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py (hunk 5 above)
Comment (xdu-chenrj) [reply dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py]: Sorry, I can't think of a better way to parse the url. I want to know your suggestion.

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/gitlab.py (hunk 4 above)
Comment (zhongjiajie) [reply dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/gitlab.py]: after we discuss `url.parse` can not handle this issue.

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py (hunk 5 above)
Comment (zhongjiajie) [reply dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py]: I mean we should use the constant variable to replace `/` and `.`, BTW, you should use

```diff
- endpoint="https://" + elements[2].split(".")[1] + ".aliyuncs.com"
+ endpoint=f"https://{elements[2].split('.')[1]}.aliyuncs.com"
```
for more pythonic

**Comment by zhongjiajie** [comment]
@xdu-chenrj It seems python lint error, could you run `tox -e auto-lint` to fix it locally and then re-submit again?

**Comment by xdu-chenrj** [comment]
> @xdu-chenrj It seems python lint error, could you run `tox -e auto-lint` to fix it locally and then re-submit again?

ok

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py (hunk 5 above)
Comment (xdu-chenrj) [reply dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/oss.py]: > ```diff
> f"https://{elements[2].split('.')[1]}.aliyuncs.com"
> ```

thanks

**Review by zhongjiajie — APPROVED** [review]
LGTM, thanks

**Pull request merged**